String.cxx

```diff
@@ -0,0 +1,184 @@
+    }
+
+     // template <class T, class = std::enable_if_t<std::is_floating_point<T>::value>>
+    // String(T val, int precision); // unimplemented
+
+    [[nodiscard]] inline auto c_str() const noexcept { return m_u.c_str(); }
+    [[nodiscard]] inline auto length() const noexcept { return m_u.length(); }
+    [[nodiscard]] inline auto charAt(unsigned idx) const noexcept { return m_u.at(idx); }
+    [[nodiscard]] inline auto& charAt(unsigned idx) noexcept { return m_u.at(idx); }
+    [[nodiscard]] inline auto operator[](unsigned idx) const noexcept { return m_u[idx]; }
+    [[nodiscard]] inline auto& operator[](unsigned idx) noexcept { return m_u[idx]; }
+
+    template <class T>[[nodiscard]] inline bool concat(const T& v) {
+        m_u += String(v).m_u;
+        return true;
+    }
+
+    inline int compareTo(const String& s) const noexcept { return std::memcmp(m_u.c_str(), s.m_u.c_str(), (std::min)(s.m_u.size(), m_u.size())); }
+
+    inline bool startsWith(const String& s) const noexcept {
+        if (s.m_u.size() > m_u.size())
+            return false;
+        return std::memcmp(m_u.c_str(), s.m_u.c_str(), s.m_u.size()) == 0;
+    }
+
+    inline bool endsWith(const String& s) const noexcept {
+        if (s.m_u.size() > m_u.size())
+            return false;
+        return std::memcmp(&*m_u.rbegin() - s.m_u.size(), s.m_u.c_str(), s.m_u.size()) == 0;
+    }
+
+    inline void getBytes(byte buffer [], unsigned length) { 
+        for (int i = 0; i < m_u.length; i++) {
+            buffer[i] = (byte)m_u[i];
+        }
+    }
+
+    inline int indexOf(const char* c) { return m_u.find(c); }
+
+    inline int indexOf(const char* c, unsigned index) { return m_u.find (c, index); }
+
+    inline int indexOf(const String& str) { return m_u.find(str); }
+
+    inline int indexOf(const String& str, unsigned index) { return m_u.find(str, index); }
+
+    inline void remove(unsigned idx) { m_u.erase(index); }
+
+    inline void remove(unsigned idx, unsigned count) { m_u.erase(idx, idx + count - 1); }
+
+    inline void replace(const String& substring1, const String& substring2) { m_u = std::regex_replace (m_u, std::regex(substring1), substring2); }
+
+    inline void reserve(unsigned size) { m_u.reserve(size); }
+
+    inline void setCharAt(unsigned index, const char* c) { m_u[index] = c; }
+
+    inline String substring(unsigned from) { return m_u.substr(from); }
+
+    inline String substring(unsigned from, unsigned to){ return m_u.substr(from, to - from); }
+
+    inline void toCharArray(const char* buffer[], unsigned length) { 
+        if (length <= m_u.length) {
+            for (int i = 0; i < length; i++) {
+                buffer[i] = m_u[i];
+            }
+        } else {
+            for ( i = 0; i < m_u.length; i++) {
+                buffer[i] = m_u[i];
+            }
+        }
+    }
+
+    inline int toInt() const noexcept {
+        if (isdigit(m_u[0]) {
+            return std::stoi(m_u);
+        } else return 0L;
```

Comment:
0L?

Reply:
The method should return a long if unsuccessful according to arduino

Reply:
+1; long suffix not needed

Reply:
> The method should return a long if unsuccessful according to arduino

Then why does the method have `int` as its return type

Reply:
From arduino:
> If no valid conversion could be performed because the String doesn’t start with a integer number, a zero is returned. Data type: long.
So change return type to long

Reply:
Mistake on my part, force of habit to make return type int when method is called toInt, thanks for noticing